Allow worker's group and development passwords.

vvereschaka opened the pull request:
This is an improvement for 'create_worker()' function, which allows
to use a single group password for a set of the workers instead of
a worker specific password.

Also it provides a default development password for the workers with
no password specified in case the buildbot instance is configured as
a development instance.

gkistanova committed:
Use pip-tools to maintain requirements.

slydiman committed:
[zorg] Avoid usage of old removed API when migrating on Buildbot 3.x.

The BuildStep.addFactoryArguments API method has been removed from Buildbot 3.x. Avoid usage of these methods.

slydiman committed:
[config] Ported notifiers initialization to Buildbot 3.x.

BB 3.x requires passing a list of builders to the notifiers through the status generators.
The notifier initialization has been updated accordingly.

Also added a new local configuration option status_email_fromaddr within the Master Options section. This option must be set with an appropriate e-mail address (llvm.buildmaster@lab.llvm.org or buildmaster@lab.llvm.org).

slydiman committed:
[zorg] Migrate LLVMInformativeMailNotifier to Buldbot 3.x.

Updated the LLVM notifier in according of changed Buildbot's API.

Based on https://github.com/gkistanova/llvm-zorg/pull/4

slydiman committed:
[zorg] Port NinjaCommand build step for Buildbot 3.x.

Port NinjaCommand build step for Buildbot 3.x.

Based on https://github.com/gkistanova/llvm-zorg/pull/2

slydiman committed:
[zorg] Port AnnotatedCommand build step for Buildbot 3.x.

Ported AnnotatedCommand for Buildbot 3.x.

Based on https://github.com/gkistanova/llvm-zorg/pull/1

slydiman committed:
Fixed the step setproperty for MSVCBuildFactory

Fixed the following exception when `env` is not defined
```
File "buildbot/master/buildbot/process/buildstep.py", line 922, in makeRemoteShellCommand
kwargs['env'] = {
builtins.TypeError: 'NoneType' object is not a mapping
```

slydiman committed:
[NFC] Replaced WithProperties with Interpolate

Not formatted yet for minimal diff.

vvereschaka committed:
Allow formatted printing for LLVMBuildFactory object.

slydiman committed:
Fixed DeprecatedApiWarning

slydiman commented on buildbot/osuosl/master/config/workers.py:
```diff
@@ -1,12 +1,59 @@
+                                      fallback=config.options.get(password_section, name, fallback=password_fallback))
+    else:
+        password = config.options.get(password_section, name, fallback=password_fallback)
+
```
Consider
```
password = config.options.get(password_section, group or "no_group", 
     fallback=config.options.get(password_section, name, fallback=password_fallback))
```

vvereschaka replied:
I think we should not support the "no_group" entry for the password section. If there is no group name then use the worker name. I.e. we don't need a "default" group name there.

slydiman replied:
`no_group` is a dummy name (must be missing in the config) to use fallback if group is None. It is a way to optimize code and remove if/else with the same `config.options.get(password_section, name, fallback=password_fallback)`. Up to you, of course.

vvereschaka replied:
got it, looks like a good idea. I'll update the changes.

vvereschaka replied:
updated

slydiman reviewed (approved):
LGTM

gkistanova committed:
[NFC] Replaced WithProperties with Interpolate after merge with the upstream.

gkistanova committed:
Use status_email_fromaddr for Reply-To as well.

vvereschaka committed:
Allow worker's group and development passwords.

This is an improvement for 'create_worker()' function, which allows
to use a single group password for a set of the workers instead of
a worker specific password.

Also it provides a default development password for the workers with
no password specified in case the buildbot instance is configured as
a development instance.

gkistanova reviewed (commented):
The patch itself looks good.

I have a concern related to this patch, though. Currently each worker gets added individually, gets staged and then promoted to production only if approved (I check the worker description, owner information, and build history before I approve the move). A worker password is a way to allow or not allow a worker into production.

With group password we are loosing a way to validate and approve individual workers, which likely would lead to missing important information in contact info and/or worker description.

Any idea how to address this and have a convenience of having group worker access?

vvereschaka commented:
@gkistanova,
>Any idea how to address this and have a convenience of having group worker access?

Hm, I though you approve the workers by approving the changes within `llvm-zorg/buildbot/osuosl/master/config/workers.py` file. Is there a case when someone adds a new worker there, you approve it, but you don't want this worker to get active?

gkistanova commented:
> Hm, I though you approve the workers by approving the changes within llvm-zorg/buildbot/osuosl/master/config/workers.py file. Is there a case when someone adds a new worker there, you approve it, but you don't want this worker to get active?

The actual procedure is:

1. A patch for a new builder and worker gets introduced for review. At the review I check if the code makes sense, free of obvious problems, adds a valuable build configuration, and such. If everything Ok, I approve.
2. Once this patch is in, I add worker/password on the staging. So, owner could connect his worker to staging and polish it, resolve issues, make it green. I do not look closely at the staged workers up to the point when owner claims it is ready for production.  At this step the workers in question could be used on staging but not in production. If owner would point such a worker to production it will not connect.
3. After the owner said it is ready, I check information configured on the worker side, such as owner contact info and worker description, as well as the build history to see if the builder is stable, reliable and reasonably fast. If everything is Ok, I add worker credentials to production and let owner know that he could move the worker there. If there are problems, owner would have to fix them before it is allowed in production.

I prefer having a control and check new workers before they would go to production. Often owners left buildbot default descriptions, put some questionable things there, or ignore issues rushing to production. 

Hope it makes sense.

vvereschaka commented:
Thank you @gkistanova .
Now I understood the workflow. It makes sense. I'll think a bit more about it
